libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Teams/Actions/SendMessagingExtensionAuthResponse.cs

```diff
@@ -0,0 +1,206 @@
+        /// Gets or sets an optional expression for the Title the Task Module response.
+        /// </summary>
+        /// <value>
+        /// An string expression. 
+        /// </value>
+        [JsonProperty("title")]
+        public ITemplate<string> Title { get; set; }
+
+        /// <summary>
+        /// Called when the dialog is started and pushed onto the dialog stack.
+        /// </summary>
+        /// <param name="dc">The <see cref="DialogContext"/> for the current turn of conversation.</param>
+        /// <param name="options">Optional, initial information to pass to the dialog.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (options is CancellationToken)
+            {
+                throw new ArgumentException($"{nameof(options)} cannot be a cancellation token");
+            }
+
+            if (this.Disabled != null && this.Disabled.GetValue(dc.State) == true)
+            {
+                return await dc.EndDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            var connectionName = ConnectionName.GetValue(dc.State);
+            if (string.IsNullOrEmpty(connectionName))
+            {
+                throw new InvalidOperationException("Messaging Extension Auth Response requires a Connection Name.");
+            }
+
+            var title = await Title.BindAsync(dc, dc.State, cancellationToken: cancellationToken).ConfigureAwait(false);
+            if (string.IsNullOrEmpty(title))
+            {
+                throw new InvalidOperationException("Messaging Extension Auth Response requires a Title.");
+            }
+
+            // When the Bot Service Auth flow completes, the action.State will contain a magic code used for verification.
+            string state = null;
+            var valueAsOjbect = dc.Context.Activity.Value as JObject;
+            if (valueAsOjbect != null)
+            {
+                state = valueAsOjbect.Value<string>("state");
+            }
+
+            string magicCode = null;
+            if (!string.IsNullOrEmpty(state))
+            {
+                int parsed = 0;
+                if (int.TryParse(state, out parsed))
+                {
+                    magicCode = parsed.ToString(CultureInfo.InvariantCulture);
+                }
+            }
+            
+            // TODO: SSO and skills token exchange
+
+            var response = await (dc.Context.Adapter as IUserTokenProvider).GetUserTokenAsync(dc.Context, connectionName, magicCode, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
>(dc.Context.Adapter as IUserTokenProvider). [](start = 33, length = 43)

not a huge fan of inlining this, I would break it out to it's own line.

Reply:
Thanks! Cleaned up SendMessagingExtensionAuthResponse some, and moved these TokenProvider calls to their own methods.  Unfortunately, this still depends on an adapter inheriting from IExtendedUserTokenProvider. This should be improved soon, with some work John is doing.